### ISSUE TITLE
reduce amount of polling

### ISSUE THREAD

**Issue by miniupnp** [atarijookie/ce-atari#137]
While working on the IKBD, I noticed there is quite a lot of polling in the ce_main_app.
Polling not only eats more CPU (power, etc) than "event driven" code, it also increases latency of processing : For example if you check for an event every 10ms, it can happen just  1ms after you last checked, and you'll wait 9ms for nothing.

IKBD was easy to fix : https://github.com/atarijookie/ce-atari/commit/7b1e88fd9f4c5c2f257674e14656254812ef7416 https://github.com/atarijookie/ce-atari/commit/a54af04a5f22b995ee7511143bb9c217d4ec01c7

I think the periodic thread can be improved a lot quite easily too.

Things will be much more difficult for GPIO as I'm not sure interrupt driven bcm2835 is available.

**Comment by mikrosk** [comment]
... not to mention it reduces number of random bugs when you mess up comparing timestamps. ;)

**Comment by atarijookie** [comment]
Holy cow, such a rework, nice work :+1: Although I hope nothing got broken :-D

That GPIO might be a problem - there is some support for some events or something when pin changes, but the description also contained some warning that there are some problems with that.... Can't remember more details on this.

**Comment by miniupnp** [comment]
timesyncThreadCode has to be improved too... polling every second instead of waiting 1 or 30 minutes (depending on the success status of the previous time sync )

**Comment by miniupnp** [comment]
@atarijookie it there a reason the TimeSync is in a dedicated thread and not in periodicThread ?

**Comment by atarijookie** [comment]
Because current implementation is very much blocking (when server doesn't respond, or something), so it would block periodicThread a lot... If anyone would improve the timeSync code to be completely asynchronous and thus non blocking, then it could be in periodicThread (although I think this is not necessary, as it doesn't cause any issue the way it is now).

**Comment by miniupnp** [comment]
well it is only eating some unecessary memory & power :)
It is very easy to implement NTP asynchronously because it is a UDP based protocol. HTTP is another problem, but it can be done in download thread, right ?

**Comment by atarijookie** [comment]
> HTTP is another problem, but it can be done in download thread, right ?

Yes, it could be changed to download it through there, and after download do the sync...

**Comment by miniupnp** [comment]
I found also that pthread_cond_wait() should be used in mounter Thread and Download thread.

**Comment by miniupnp** [comment]
Downloader could be improved to use CURL asynchronous API too :
https://curl.haxx.se/libcurl/c/libcurl-multi.html

**Comment by miniupnp** [comment]
Looks like I forgot https://github.com/atarijookie/ce-atari/blob/master/ce_main_app/floppy/imagesilo.cpp#L49

**Comment by miniupnp** [comment]
the current code check for Hans & Franz "ATN" by polling pin 16 and 18 (BCM 23 & 24) of the GPIO.
```
bool spi_atn(int whichSpiAtn)
{
    BYTE val;
	
    val = bcm2835_gpio_lev(whichSpiAtn); 
	
    return (val == HIGH);  // returns true if pin is high, returns false if pin is low
}
```
It would be better to use "interrupt" driven code. It seems possible using the GPIOsysfs API  : 
https://www.raspberrypi.org/forums/viewtopic.php?t=119472
```
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/time.h>
#include <poll.h>

#define GPIO 4

int main(int argc, char *argv[])
{
   char str[256];
   struct timeval tv;
   struct pollfd pfd;
   int fd, gpio;
   char buf[8];

   /*
      Prior calls assumed.
      sudo sh -c "echo 4      >/sys/class/gpio/export"
      sudo sh -c "echo in     >/sys/class/gpio/gpio4/direction"
      sudo sh -c "echo rising >/sys/class/gpio/gpio4/edge"
   */

   if (argc > 1) gpio = atoi(argv[1]);
   else          gpio = GPIO;

   sprintf(str, "/sys/class/gpio/gpio%d/value", gpio);

   if ((fd = open(str, O_RDONLY)) < 0)
   {
      fprintf(stderr, "Failed, gpio %d not exported.\n", gpio);
      exit(1);
   }

   pfd.fd = fd;

   pfd.events = POLLPRI;

   lseek(fd, 0, SEEK_SET);    /* consume any prior interrupt */
   read(fd, buf, sizeof buf);

   poll(&pfd, 1, -1);         /* wait for interrupt */

   lseek(fd, 0, SEEK_SET);    /* consume interrupt */
   read(fd, buf, sizeof buf);

   exit(0);
}
```

**Comment by atarijookie** [comment]
That would be awesome, but it will require massive amount of testing - lots of reading and writing over ACSI / SCSI to see if this doesn't screw up somehow... (e.g. false interrupts)

**Comment by miniupnp** [comment]
missed interrupts is a bigger problem :)
You have a kind of "synchro" over SPI after getting the ATN.
(waiting for "0xcafe")

**Comment by miniupnp** [comment]
Hum my first tests are not conclusive :(
@atarijookie  I'm using /sys/class/gpio/gpio23/ /sys/class/gpio/gpio24/  (SPI_ATN_HANS / SPI_ATN_FRANZ)
Are they the right GPIO ?

**Comment by miniupnp** [comment]
it is working...

**Comment by miniupnp** [comment]
there are strange things about floppy encoding, I think I need to clean the code, right ? :)
the inter-thread communications are a bit strange :)

**Comment by atarijookie** [comment]
> there are strange things about floppy encoding, I think I need to clean the code, right ? :)

Yeah, why not...

> the inter-thread communications are a bit strange :)

It's because I'm retarded, that's why ;)

**Comment by miniupnp** [comment]
hum I have to rebase my code for interrupt driven GPIO stuff
